azavea/pfb-network-connectivity
# Refactor simplify_geom method to preserve topology if simplification issues

## Overview

_Hopefully_ improves the failure rate, and is specifically intended to avoid geometries exhibiting the issue noted in #452.

It's difficult to locally test all of the geometries that are on production to ensure there aren't any signficant regressions, happy to hear any ideas you might have for that.

### Demo

Here is the Ann Arbor, MI shapefile uploaded to my local box with the changes in the PR applied. The geojson was retrieved via:

http://localhost:9200/api/neighborhoods_bounds_geojson/b578299c-effb-4bce-8a0a-ed3f6d04dceb/?format=json

Which is the same endpoint that the client app uses to show boundaries on the map.

Putting the result into geojson.io gives us:

![screen shot 2018-04-10 at 11 34 46 am](https://user-images.githubusercontent.com/1818302/38567108-57996fca-3cb3-11e8-8fbd-b5b081eec14f.png)

Where the simple geometry is ~500 points down from the original ~1800.

### Notes

I tested this on 5-6 different known boundaries that we've uploaded in the past. It appears robust at a glance to empty, valid and non-polygon geometries, but I'm sure there's some failure case we're not anticipating that it doesn't handle well. 

I added a few simple tests that broadly verify that simplification was successful for a boundary that used to work with the old simplify_geom method, and a boundary that didn't work but now does that exhibits the preserve_topology issue. Happy to expand there a bit if we think that's appropriate.

I'm unsure 

## Testing Instructions

- `./scripts/server`
- Via the admin interface at http://localhost:9301/#/admin/neighborhoods/ upload a variety of new neighborhoods via shapefile. Ensure they all upload and simplify correctly by verifying the json output of `http://localhost:9200/api/neighborhoods_bounds_geojson/:neighborhoodId/?format=json` in a viewer such as geojson.io. Good choices that exhibit the issue in #488 are:
  - https://s3.amazonaws.com/production-pfb-storage-us-east-1/neighborhood_boundaries/ann-arbor-mi/ann-arbor-mi.zip
  - https://s3.amazonaws.com/production-pfb-storage-us-east-1/neighborhood_boundaries/austin-tx/austin-tx.zip
  - https://s3.amazonaws.com/production-pfb-storage-us-east-1/neighborhood_boundaries/denver/denver.zip
- The new tests should also pass.


Closes #488
